parquet-variant/src/main/java/org/apache/parquet/variant/VariantElementConverter.java

```diff
@@ -0,0 +1,668 @@
+            case MICROS:
+              typedConverter = new VariantTimestampConverter();
+              break;
+            case NANOS:
+              typedConverter = new VariantTimestampNanosConverter();
+              break;
+          }
+        } else {
+          switch (timestampType.getUnit()) {
+            case MILLIS:
+              throw new UnsupportedOperationException("MILLIS not supported by Variant");
+            case MICROS:
+              typedConverter = new VariantTimestampNtzConverter();
+              break;
+            case NANOS:
+              typedConverter = new VariantTimestampNanosNtzConverter();
+              break;
+          }
+        }
+      } else if (annotation instanceof LogicalTypeAnnotation.TimeLogicalTypeAnnotation) {
+        LogicalTypeAnnotation.TimeLogicalTypeAnnotation timeType =
+            (LogicalTypeAnnotation.TimeLogicalTypeAnnotation) annotation;
+        if (timeType.isAdjustedToUTC() || timeType.getUnit() != MICROS) {
+          throw new UnsupportedOperationException(timeType + " not supported by Variant");
+        } else {
+          typedConverter = new VariantTimeConverter();
+        }
+      } else if (annotation instanceof LogicalTypeAnnotation.UUIDLogicalTypeAnnotation) {
+        typedConverter = new VariantUuidConverter();
+      } else if (annotation instanceof LogicalTypeAnnotation.StringLogicalTypeAnnotation) {
+        typedConverter = new VariantStringConverter();
+      } else if (primitiveType == BINARY) {
+        typedConverter = new VariantBinaryConverter();
+      } else {
+        throw new UnsupportedOperationException("Unsupported shredded value type: " + primitive);
+      }
+      return typedConverter;
+    }
+  }
+
+  static class VariantStringConverter extends VariantScalarConverter {
+    @Override
+    public void addBinary(Binary value) {
+      builder.builder.appendString(value.toStringUsingUTF8());
+    }
+  }
+
+  static class VariantBinaryConverter extends VariantScalarConverter {
+    @Override
+    public void addBinary(Binary value) {
+      builder.builder.appendBinary(value.toByteBuffer());
+    }
+  }
+
+  static class VariantDecimalConverter extends VariantScalarConverter {
+    private int scale;
+
+    VariantDecimalConverter(int scale) {
+      super();
+      this.scale = scale;
+    }
+
+    @Override
+    public void addBinary(Binary value) {
+      builder.builder.appendDecimal(new BigDecimal(new BigInteger(value.getBytes()), scale));
+    }
+
+    @Override
+    public void addLong(long value) {
+      BigDecimal decimal = BigDecimal.valueOf(value, scale);
+      builder.builder.appendDecimal(decimal);
+    }
+
+    @Override
+    public void addInt(int value) {
+      BigDecimal decimal = BigDecimal.valueOf(value, scale);
+      builder.builder.appendDecimal(decimal);
+    }
+  }
+
+  static class VariantUuidConverter extends VariantScalarConverter {
```

Comment:
UUID is usually a special case where we don't downcase in method or class names. I think this should be `VariantUUIDConverter` instead of `VariantUuidConverter`.